src/blend-impl.ts

```diff
@@ -2,53 +2,80 @@ import {
   abs,
   add,
   div,
-  float,
+  FLOAT0,
+  FLOAT05,
+  FLOAT1,
+  FLOAT2,
   gte,
   lt,
   lte,
   max,
   min,
   mul,
   sub,
+  Term,
   ternary,
   vec3,
+  Vec3Term,
   vec4,
+  Vec4Term,
 } from "@thi.ng/shader-ast";
-import { BlendModeFloat, BlendModeVec } from "./api";
+import type {
+  BlendModeFloat,
+  BlendModeVec3,
+  BlendModeVec4,
+  ColorTerm,
+} from "./api";
 
-export const blendAddVec3: BlendModeVec<"vec3"> = (base, blend) =>
-  min(add(base, blend), vec3(1.0));
+export const blendAddVec3: BlendModeVec3 = (base, blend) =>
+  min(add(base, blend), vec3(FLOAT1));
 
-export const blendAddVec4: BlendModeVec<"vec4"> = (base, blend) =>
-  min(add(base, blend), vec4(1.0));
+export const blendAddVec4: BlendModeVec4 = (base, blend) =>
+  min(add(base, blend), vec4(FLOAT1));
 
-export const blendAverageVec: BlendModeVec<"vec3" | "vec4"> = (base, blend) =>
-  div(add(base, blend), float(2.0));
+export function blendAverageVec(base: Vec3Term, blend: Vec3Term): Vec3Term;
+export function blendAverageVec(base: Vec4Term, blend: Vec4Term): Vec4Term;
+export function blendAverageVec(base: ColorTerm, blend: ColorTerm): Term<any> {
```

Comment:
Unfortunately we need those function overrides to ensure these functions are always called w/ same type for both args...

Reply:
Yeah i had thought about overloads, but now I see why this is a requirement!

Reply:
Those overloads are only needed because the way TS handles union types in generics... it does make sense as it is, but I wish there'd be a better way to express if `<T>` is matched from a type union, then `T` is the same match everywhere, rather than just the union itself (and thus allowing the other options from the union too)... anyhow, maybe a bit overkill, but at least you've seen the usage pattern now... :)